### PR TITLE
Add support for sourcemap

### LINKED ISSUE #4: SourceMap broken
When using this loader on TypeScript, sourcemaps are broken making debuging impossible.

### PR DESCRIPTION
**Pull request by Toilal**
Close #4

Well, I dive deep into sourcemap generation to understand how to implement this, bit it seems to be the right way to support sourcemap generated by any previous loader in the chain. 

Tested with TypeScript.

### CODE REVIEW

**Comment by Toilal** [comment]
I'll fix build and make more manual tests tomorrow.

**Comment by Toilal** [comment]
Build fixed, and works on TypeScript and ES6 using #inline-source-map.

File: lib/index.js [hunk 1]
```diff
@@ -3,12 +3,18 @@ const estraverse = require('estraverse')
 const genCode = require('escodegen')
 const genId = require('./gen-id')
 
-module.exports = function(contents) {
+const sourceMap = require('source-map')
+const SourceMapConsumer = sourceMap.SourceMapConsumer
+const SourceMapGenerator = sourceMap.SourceMapGenerator
```
Comment (ktsn) [inline lib/index.js]: Can we remove this line since it's not used?

File: lib/index.js [hunk 2]
```diff
@@ -39,8 +45,30 @@ module.exports = function(contents) {
     }
   })
 
-  return genCode.generate(res)
-    + (hasComponent ? genHotReload(id) : '')
+  if (hasComponent) {
```
Comment (ktsn) [inline lib/index.js]: I prefer to negate this condition and `return` in this block early so that we can reduce the nesting level 🙂

**Review by ktsn — CHANGES_REQUESTED** [review]
I think we should consider one more case - when previous source map does not exist even though the user enables source map. (e.g. there are no loader before this loader)

We can use `this.sourceMap` whether we should output source map.
https://webpack.js.org/api/loaders/#sourcemap

**Comment by Toilal** [comment]
Should be ok now. Thanks for the really good review !

File: lib/index.js [hunk 3]
```diff
@@ -3,12 +3,17 @@ const estraverse = require('estraverse')
...
+const sourceMap = require('source-map')
+const SourceMapConsumer = sourceMap.SourceMapConsumer
+
+module.exports = function(contents, sourcemap) {
   const id = genId(this.resourcePath)
   let hasComponent = false
 
   const ast = acorn.parse(contents, {
-    sourceType: 'module'
+    sourceType: 'module',
```
Comment (Toilal) [inline lib/index.js]: I also think I can remove locations and sourceFile option when this.sourceMap is falsy.

**Comment by ktsn** [comment]
@Toilal Looks great. Can we fix the lint error? 🙂

**Comment by Toilal** [comment]
Fixed

**Review by ktsn — APPROVED** [review]

**Comment by ktsn** [comment]
Thanks. Great work! 👍

**Pull request merged**